### PR TITLE
Inherited optional labels check

### PR DESCRIPTION
**Pull request by rcerven**

### CODE REVIEW

**Comment by lachmanfrantisek** [comment]
[test]

File: colin/cli/colin.py [hunk 1]
```diff
@@ -21,7 +21,8 @@
 import click
 import six
 
-from colin.utils.cmd_tools import get_version_msg_from_the_cmd, is_rpm_installed
+from colin.utils.cmd_tools import get_version_msg_from_the_cmd, is_rpm_installed, \
+    get_version_of_the_python_package
```
Comment (lslebodn) [inline colin/cli/colin.py]: May I know where do we need `get_version_of_the_python_package`?

I cannot used it anywhere. But maybe I opverlooked something.

File: colin/cli/colin.py [hunk 2]
```diff
@@ -48,6 +49,7 @@ def cli():
 @click.command(name="check",
                context_settings=CONTEXT_SETTINGS)
 @click.argument('TARGET', type=click.STRING)
+@click.argument('parent_target', type=click.STRING, required=False)
```
Comment (lslebodn) [inline colin/cli/colin.py]: IIRC we agreed to pass output of skopeo inspect  (which is in json)
Do we expect that `parent_target` will contain path to json file or string with json?

I do not have any preferences. I can pass anything to colin :-).
I just cannot see conversion string(json) -> object in PR.
But I check RP very briefly.

Comment (lachmanfrantisek) [reply colin/cli/colin.py]: If I understand that correctly, this is a different approach -- we have a name of the parent and handle it in the same way, as we are handling the "main" target.

I am not against this. It can probably have some overhead with handling the parent image, but it is more general so we can create other checks with that as well.

I also have no problem with getting info from JSON.

File: colin/core/checks/labels.py [hunk 3]
```diff
@@ -73,3 +75,43 @@ def check(self, target):
...
+            'io.k8s.display-name',
+            'io.openshift.tags',
+        ]
+        passed = True
+
+        if target.parent_target:
+            for label in optional_labels:
+                if label in target.labels:
+                    if label in target.parent_target.labels:
+                        if target.labels[label] == target.parent_target.labels[label]:
```
Comment (lachmanfrantisek) [inline colin/core/checks/labels.py]: It took me a while to see what's going on. What about using set operations to make it clear?

```python
        if target.parent_target:
            labels_to_check = set(optional_labels) - set(target_labels) - set(target.parent_target.labels)
            for label in labels_to_check:
                if target.labels[label] == target.parent_target.labels[label]:
```

**Comment by lachmanfrantisek** [comment]
And please, fix the `test_labels_in_image` test:

```
22:36:36     def test_labels_in_image(target_label):
22:36:36         result = get_results_from_colin_labels_image(target=target_label)
22:36:36         assert result
22:36:36         expected_dict = {"maintainer_label": "PASS",
22:36:36                          "name_label": "PASS",
22:36:36                          "com.redhat.component_label": "PASS",
22:36:36                          "summary_label": "PASS",
22:36:36                          "version_label": "PASS",
22:36:36                          "run_or_usage_label": "PASS",
22:36:36                          "release_label": "FAIL",
22:36:36                          "architecture_label": "FAIL",
22:36:36                          "url_label": "PASS",
22:36:36                          "help_label": "FAIL",
22:36:36                          "build-date_label": "FAIL",
22:36:36                          "distribution-scope_label": "FAIL",
22:36:36                          "vcs-ref_label": "FAIL",
22:36:36                          "vcs-type_label": "FAIL",
22:36:36                          "description_label": "PASS",
22:36:36                          "io.k8s.description_label": "PASS",
22:36:36                          "vcs-url_label": "FAIL",
22:36:36                          "help_file_or_readme": "FAIL",
22:36:36                          # "cmd_or_entrypoint": "PASS",
22:36:36                          # "no_root": "FAIL",
22:36:36                          }
22:36:36     
22:36:36         if isinstance(target_label, DockerfileTarget):
22:36:36             expected_dict.update({'description_or_io.k8s.description_label': 'PASS',
22:36:36                                   'from_tag_not_latest': 'FAIL',
22:36:36                                   'maintainer_deprecated': 'PASS'}
22:36:36                                  )
22:36:36             del (expected_dict['help_file_or_readme'])
22:36:36             del (expected_dict['run_or_usage_label'])
22:36:36         labels_dict = {}
22:36:36         for res in result.results:
22:36:36             labels_dict[res.check_name] = res.status
22:36:36 >       assert labels_dict == expected_dict
22:36:36 E       AssertionError: assert {'architectur...: 'PASS', ...} == {'architecture...: 'PASS', ...}
22:36:36 E         Common items:
22:36:36 E         {'architecture_label': 'FAIL',
22:36:36 E          'build-date_label': 'FAIL',
22:36:36 E          'com.redhat.component_label': 'PASS',
22:36:36 E          'description_label': 'PASS',
22:36:36 E          'distribution-scope_label': 'FAIL',
22:36:36 E          'help_file_or_readme': 'FAIL',
22:36:36 E          'help_label': 'FAIL',
22:36:36 E          'io.k8s.description_label': 'PASS',
22:36:36 E          'maintainer_label': 'PASS',
22:36:36 E          'name_label': 'PASS',
22:36:36 E          'release_label': 'FAIL',
22:36:36 E          'run_or_usage_label': 'PASS',
22:36:36 E          'summary_label': 'PASS',
22:36:36 E          'url_label': 'PASS',
22:36:36 E          'vcs-ref_label': 'FAIL',
22:36:36 E          'vcs-type_label': 'FAIL',
22:36:36 E          'vcs-url_label': 'FAIL',
22:36:36 E          'version_label': 'PASS'}
22:36:36 E         Left contains more items:
22:36:36 E         {'inherited_labels': 'PASS'}
22:36:36 E         Full diff:
22:36:36 E         {'architecture_label': 'FAIL',
22:36:36 E         'build-date_label': 'FAIL',
22:36:36 E         'com.redhat.component_label': 'PASS',
22:36:36 E         'description_label': 'PASS',
22:36:36 E         'distribution-scope_label': 'FAIL',
22:36:36 E         'help_file_or_readme': 'FAIL',
22:36:36 E         'help_label': 'FAIL',
22:36:36 E         -  'inherited_labels': 'PASS',
22:36:36 E         'io.k8s.description_label': 'PASS',
22:36:36 E         'maintainer_label': 'PASS',
22:36:36 E         'name_label': 'PASS',
22:36:36 E         'release_label': 'FAIL',
22:36:36 E         'run_or_usage_label': 'PASS',
22:36:36 E         'summary_label': 'PASS',
22:36:36 E         'url_label': 'PASS',
22:36:36 E         'vcs-ref_label': 'FAIL',
22:36:36 E         'vcs-type_label': 'FAIL',
22:36:36 E         'vcs-url_label': 'FAIL',
22:36:36 E         'version_label': 'PASS'}
22:36:36 
```

(https://ci.centos.org/job/user-cont-colin-pr/101/console)

File: colin/cli/colin.py (hunk 2 above)
Comment (lslebodn) [reply colin/cli/colin.py]: > If I understand that correctly, this is a different approach -- we have a name of the parent and handle it in the same way, as we are handling the "main" target.

I am fine with that but it really need to be handled the same way as `target` in `class OstreeTarget`.
And I cannot see any validation for `parent target there`. lines 328 ..335

In another words, it has to work also with ostree format

e.g.
```
colin --target-type=ostree --parent-target ostree:${parentImage}@/var/tmp/os3/ ostree:${imageToTest}@/var/tmp/os3/
```

File: colin/cli/colin.py (hunk 1 above)
Comment (rcerven) [reply colin/cli/colin.py]: will delete it, got it from cherry picking

File: colin/cli/colin.py (hunk 2 above)
Comment (rcerven) [reply colin/cli/colin.py]: it works for both ImageTarget and OstreeTarget

both of them are now accepting parent_target

File: colin/core/checks/labels.py (hunk 3 above)
Comment (rcerven) [reply colin/core/checks/labels.py]: sure I can use set operation

but we will need intersection not difference

Comment (lachmanfrantisek) [reply colin/core/checks/labels.py]: Yes, you're right...;-)

**Comment by TomasTomecek** [comment]
[test]

File: colin/core/checks/labels.py [hunk 4]
```diff
@@ -73,3 +77,43 @@ def check(self, target):
...
+            'summary',
+            'description',
+            'io.k8s.description',
+            'io.k8s.display-name',
+            'io.openshift.tags',
+        ]
+        passed = True
+
+        if target.parent_target:
+            labels_to_check = (set(optional_labels) & set(target.labels)
```
Comment (TomasTomecek) [inline colin/core/checks/labels.py]: I don't follow why the variable is named optional when it seems they are required actually since you are doing `optional AND child_labels AND parent_labels`

File: colin/core/checks/labels.py [hunk 5]
```diff
@@ -73,3 +77,43 @@ def check(self, target):
...
+        :param message: str
+        :param description: str
+        :param reference_url: str
+        :param tags: [str]
+        """
+        super(InheritedOptionalLabelAbstractCheck, self) \
+            .__init__(message, description, reference_url, tags)
+
+    def check(self, target):
+        optional_labels = [
```
Comment (TomasTomecek) [inline colin/core/checks/labels.py]: I'm also wondering if we want to hardcode this in the codebase; wouldn't it be better to store it in a ruleset?

File: colin/core/checks/labels.py (hunk 4 above)
Comment (rcerven) [reply colin/core/checks/labels.py]: well those are optional labels, but we first need to collect all which are present in both image and parent image, and then we compare values

File: colin/core/checks/labels.py (hunk 5 above)
Comment (rcerven) [reply colin/core/checks/labels.py]: reading it now from ruleset

**Comment by rcerven** [comment]
[test]

**Comment by TomasTomecek** [comment]
[test]

**Comment by TomasTomecek** [comment]
You need to be on jenkins' whitelist in order for your changes to be tested. Wanna commit rights? Planning on sending more code?

**Comment by rcerven** [comment]
@TomasTomecek ic, no problem, I have just this PR nothing else planned

**Comment by rcerven** [comment]
@TomasTomecek I've mentioned that argument in here : https://github.com/user-cont/colin/pull/230/files#diff-a2d1c7f29c5c130f37eea40bf6252b0cR228

**Comment by TomasTomecek** [comment]
[test]

File: colin/checks/labels.fmf [hunk 6]
```diff
@@ -221,3 +221,10 @@ tags: ["label"]
   labels: ["version"]
   required: True
   value_regex: Null
+
+/inherited_labels:
+  class: InheritedOptionalLabelCheck
+  message: "Optional labels are only inherited"
+  description: "Check if optional labels provided with 'labels_list' argument, are only inherited"
+  reference_url: "?????"
```
Comment (lachmanfrantisek) [inline colin/checks/labels.fmf]: It would be nice to have some URL with more information, do we have any?

**Comment by lslebodn** [comment]
@rcerven I am not sure whether it works as expected.

Let say we have two images:
a) registry.access.redhat.com/rhel7/rsyslog:7.6-20 (TARGET)
b) docker.io/fedora:29 (PARENT_TARGET)

rsyslog image has all labels which we are interested in but fedora does not have

It works as expected if fedora is set as parent image
```
[test@ac42a71d9ace ~]$ colin -r fedora --target-type=ostree ostree:registry.access.redhat.com/rhel7/rsyslog:7.6-20@/var/tmp/os3/ ostree:docker.io/fedora:29@/var/tmp/os3/
PASS:Label 'architecture' has to be specified.
PASS:Label 'build-date' has to be specified.
PASS:Label 'description' has to be specified.
PASS:Label 'distribution-scope' has to be specified.
FAIL:Label 'help' has to be specified.
PASS:Label 'io.k8s.description' has to be specified.
PASS:Label 'url' has to be specified.
PASS:Label 'vcs-ref' has to be specified.
PASS:Label 'vcs-type' has to be specified.
FAIL:Label 'vcs-url' has to be specified.
PASS:Label 'com.redhat.component' has to be specified.
PASS:Label 'maintainer' has to be specified.
PASS:Label 'name' has to be specified.
PASS:Label 'release' has to be specified.
PASS:Label 'summary' has to be specified.
PASS:Label 'version' has to be specified.
FAIL:The 'helpfile' has to be provided.
PASS:Label 'usage' has to be specified.
PASS:Optional labels are only inherited
```

But it should fail if I change order (rsyslog is parent)
```
[test@ac42a71d9ace ~]$ colin -r fedora --target-type=ostree ostree:docker.io/fedora:29@/var/tmp/os3/ ostree:registry.access.redhat.com/rhel7/rsyslog:7.6-20@/var/tmp/os3/
FAIL:Label 'architecture' has to be specified.
FAIL:Label 'build-date' has to be specified.
FAIL:Label 'description' has to be specified.
FAIL:Label 'distribution-scope' has to be specified.
FAIL:Label 'help' has to be specified.
FAIL:Label 'io.k8s.description' has to be specified.
FAIL:Label 'url' has to be specified.
FAIL:Label 'vcs-ref' has to be specified.
FAIL:Label 'vcs-type' has to be specified.
FAIL:Label 'vcs-url' has to be specified.
FAIL:Label 'com.redhat.component' has to be specified.
PASS:Label 'maintainer' has to be specified.
FAIL:Label 'name' has to be specified.
FAIL:Label 'release' has to be specified.
FAIL:Label 'summary' has to be specified.
FAIL:Label 'version' has to be specified.
FAIL:The 'helpfile' has to be provided.
FAIL:Label 'usage' has to be specified.
PASS:Optional labels are only inherited
```

**Comment by lslebodn** [comment]
@rcerven my previous expectation is based on following requirement:

> The problem is that if a base image defines a label and a layered image does not override it, it is inherited. That's not what we want if the layered image is a different product.

Or another definition:
> Fail build if optional labels are only inherited

**Comment by lslebodn** [comment]
Maybe previous situation cannot happen in OSBS but we should not rely on such situation IMHO.

Another issue is missing error reporting:
e.g. base image `registry.access.redhat.com/rhel7:7.6` define label `io.openshift.tags` which is inherited by rsyslog image. But we cannot anything in json output

```
[test@9d9eb4a052c0 ~]$ colin -r fedora --target-type=ostree ostree:registry.access.redhat.com/rhel7/rsyslog:7.6-20@/var/tmp/os3/ ostree:registry.access.redhat.com/rhel7:7.6@/var/tmp/os3/ --json test.json
PASS:Label 'architecture' has to be specified.
PASS:Label 'build-date' has to be specified.
PASS:Label 'description' has to be specified.
PASS:Label 'distribution-scope' has to be specified.
FAIL:Label 'help' has to be specified.
PASS:Label 'io.k8s.description' has to be specified.
PASS:Label 'url' has to be specified.
PASS:Label 'vcs-ref' has to be specified.
PASS:Label 'vcs-type' has to be specified.
FAIL:Label 'vcs-url' has to be specified.
PASS:Label 'com.redhat.component' has to be specified.
PASS:Label 'maintainer' has to be specified.
PASS:Label 'name' has to be specified.
PASS:Label 'release' has to be specified.
PASS:Label 'summary' has to be specified.
PASS:Label 'version' has to be specified.
FAIL:The 'helpfile' has to be provided.
PASS:Label 'usage' has to be specified.
FAIL:Optional labels are only inherited

PASS:15 FAIL:4 
[test@9d9eb4a052c0 ~]$ tail -n 14 test.json
            "reference_url": "https://fedoraproject.org/wiki/Container:Guidelines#LABELS",
            "logs": []
        },
        {
            "name": "inherited_labels",
            "ok": false,
            "status": "FAIL",
            "description": "Check if optional labels provided with 'labels_list' argument, are only inherited",
            "message": "Optional labels are only inherited",
            "reference_url": "?????",
            "logs": []
        }
    ]
}
```

It is only in debug output.

File: colin/cli/colin.py [hunk 7]
```diff
@@ -48,6 +48,7 @@ def cli():
 @click.command(name="check",
                context_settings=CONTEXT_SETTINGS)
 @click.argument('TARGET', type=click.STRING)
+@click.argument('PARENT_TARGET', type=click.STRING, required=False)
```
Comment (lslebodn) [inline colin/cli/colin.py]: It would be better if we change CLI interface a little bit.

This version relies on optional positional parameter
```
sh# colin checks -h
Usage: colin checks [OPTIONS] TARGET [PARENT_TARGET]

  Check the image/dockerfile (default).
```

But optional parameters should not be positional. Named parameter fits such use-case better IMHO
e.g. --parent-target

@TomasTomecek @lachmanfrantisek
What do you think ?

**Comment by rcerven** [comment]
@lslebodn  I am not sure why you would expect it to fail when you switch order of

registry.access.redhat.com/rhel7/rsyslog:7.6-20 (TARGET)
docker.io/fedora:29 (PARENT_TARGET)

fedora image doesn't have any of those labels, so that means that none of those labels were inherited

File: colin/checks/labels.fmf (hunk 6 above)
Comment (rcerven) [reply colin/checks/labels.fmf]: maybe rcm might have something, but I'd say that will be internal

**Comment by rcerven** [comment]
as for error reporting maybe I can add those debug messages as well to the 'logs' in that result, unless you have some other suggestion

**Comment by rcerven** [comment]
I've added error reporting to logs as well

**Comment by lslebodn** [comment]
> @lslebodn I am not sure why you would expect it to fail when you switch order of
>
> registry.access.redhat.com/rhel7/rsyslog:7.6-20 (PARENT_TARGET)
> docker.io/fedora:29 (TARGET)
> 

I meant other way. That labels are neither inherited not overridden in TARGET.
(becase labels are not present in TARGET)
Is it a problem/(valid test-case)?

And another corner case. Do we need to care about case where TARGET and PARENT_TARGET
have the same value for label in Dockerfile ?
(label was not technically inherited by OSBS but  this check will fail).
Such situation is not possible to detect from the image target. 
Is it a problem/(valid use-case)?

**Comment by rcerven** [comment]
@lslebodn if those labels aren't present in the image it doesn't matter since they are "optional",
what we are trying to catch are the same labels in image and it's parent

when you read original jira story, we moved from that Dockerfile parsing, just to image comparison

**Comment by lslebodn** [comment]
>  @lslebodn if those labels aren't present in the image it doesn't matter since they are "optional", what we are trying to catch are the same labels in image and it's parent

I just wanted to be sure :-)
Thank you very much for clarification.

File: colin/cli/colin.py (hunk 7 above)
Comment (TomasTomecek) [reply colin/cli/colin.py]: Yup, sounds good.

Comment (rcerven) [reply colin/cli/colin.py]: changed

**Comment by rcerven** [comment]
using now optional parameter  --parent-target

**Comment by lslebodn** [comment]
> using now optional parameter --parent-target

Thank you. It works as expected.
I think we can remove WIP :-) and merge PR

We can update `reference_url` later if we manage to have some document.
